File: Pythonleetcode/Leetcode138.py
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution(object):
    def copyRandomList(self, head):
        if not head:
            return None
        dict={}
        ans=Node(0)
        cur_ans=ans
        cur_head=head
        dict[cur_head]=cur_ans

        while cur_head:
            cur_ans.val=cur_head.val
            if cur_head.next:
                cur_ans.next=Node(0)
            cur_head=cur_head.next
            cur_ans=cur_ans.next
            dict[cur_head]=cur_ans

        cur_ans=ans
        cur_head=head

        while cur_head:
            cur_ans.random=dict[cur_head.random]
            cur_head=cur_head.next
            cur_ans=cur_ans.next


        return ans

File: CppPractice/CppLeetcode/Leetcode138.cpp

// Definition for a Node.
// class Node {
// public:
//     int val;
//     Node* next;
//     Node* random;
    
//     Node(int _val) {
//         val = _val;
//         next = NULL;
//         random = NULL;
//     }
// };


class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*,Node*> theMap;
        Node* cur = head;
        while(cur)
        {
            theMap[cur] = new Node(cur->val);
            cur = cur->next;
        }
        cur =head;
        while(cur)
        {
            theMap[cur]->next = theMap[cur->next];
            theMap[cur]->random = theMap[cur->random];
            cur=cur->next;
        }
        return theMap[head];
    }
};